### ISSUE TITLE
Upgrading from babel-core 5.3.3 to 5.4.5 breaks everything

### ISSUE THREAD

**Issue by psbanka** [babel/babel#1596]
Sorry for the lack of specificity of this message, but we upgraded from `babel-core@5.3.3` / `babel-loader@5.0.0` to the latest `babel-loader@5.0.0` / `babel-core@5.4.5` this morning and we are encountering major breaking changes our code. Pinning the versions back to `5.0.0`/`5.3.3` has resolved them.

Do you have any advice for how to address these breaking changes, or is this a mistake in this particular release?

Thanks!

**Comment by sebmck** [comment]
I cannot debug anything with so little information. Do you have a stack trace or a reproducible example? There are no breaking changes as far as I know.

**Comment by psbanka** [comment]
Ah, specifically this is a JSX issue. Apparently when assigning props such as `foo={foo}` within a JSX property, we get this issue.

https://github.com/cyaninc/beaker/pull/55

**Comment by sandersky** [comment]
The following code
```js
let myVarName = 'something cool';
```

is getting transcompiled to something lie:
```js
let _myVarName2 = 'something cool';
```

**Comment by sebmck** [comment]
@sandersky [No it doesn't.](http://babeljs.io/repl/#?experimental=true&evaluate=true&loose=false&spec=false&code=let%20myVarName%20%3D%20'something%20cool'%3B&playground=false) Do you have a reproducible example in the REPL?

**Comment by job13er** [comment]
Sorry, I'm working on the reproducible example, just trying to get our team un-blocked with a workaround first.

**Comment by job13er** [comment]
@sebmck Sorry. I can't get it to reproduce using just `babel` it may be an issue with `webpack` and how it's handling the transpiled code. 

Somehow

```js
for (let i = 0, len = this.props.menuList.length; i < len; i++) {
    const menuItem = this.props.menuList[i];
    const enabled = this.props.enabled[menuItem.idName];
    const hidden = i >= hiddenIndex;
    const defaultActive = (i === this.props.initialActive) ? true : false;

    if (menuItem.subMenus && !menuItem && menuItem) {
        navItems.push(
            <NavDropDown
                id={menuItem.linkName}
                label={menuItem.name}
                menuItems={menuItem.subMenus}
            />
        );
    } else {
        navItems.push(
            <NavItemClass
                defaultActive={defaultActive}
                enabled={enabled}
                hidden={hidden}
                key={menuItem.idName}
                menuItem={menuItem}
            />
        );
    }

}
```

Is landing in the browser as 

```js
for (var i = 0, len = this.props.menuList.length; i < len; i++) {
    var _menuItem3 = this.props.menuList[i];
    var _enabled = this.props.enabled[_menuItem3.idName];
    var _hidden = i >= hiddenIndex;
    var _defaultActive = i === this.props.initialActive ? true : false;

    if (_menuItem3.subMenus && !_menuItem3 && _menuItem3) {
        navItems.push(React.createElement(NavDropDown, {
            id: _menuItem3.linkName,
            label: _menuItem3.name,
            menuItems: _menuItem3.subMenus
        }));
    } else {
        navItems.push(React.createElement(NavItemClass, {
            _defaultActive: _defaultActive,
            _enabled: _enabled,
            _hidden: _hidden,
            key: _menuItem3.idName,
            _menuItem3: _menuItem3
        }));
    }
}
```

Even though the Try It Out shows babel spitting out:

```js
for (var i = 0, len = this.props.menuList.length; i < len; i++) {
    var menuItem = this.props.menuList[i];
    var enabled = this.props.enabled[menuItem.idName];
    var hidden = i >= hiddenIndex;
    var defaultActive = i === this.props.initialActive ? true : false;

    if (menuItem.subMenus && !menuItem && menuItem) {
        navItems.push(React.createElement(NavDropDown, {
            id: menuItem.linkName,
            label: menuItem.name,
            menuItems: menuItem.subMenus
        }));
    } else {
        navItems.push(React.createElement(NavItemClass, {
            defaultActive: defaultActive,
            enabled: enabled,
            hidden: hidden,
            key: menuItem.idName,
            menuItem: menuItem
        }));
    }
}
```

Even when I include the entire file.

**Comment by sebmck** [comment]
Can you post the entire file? A snippet wont accurately show the issue as the context is extremely important. (ie. if it's in a function etc)

**Comment by job13er** [comment]
OK, running the command-line `babel` does not yield the same as your Try It Out page.
Yes. I'll include the whole thing. Now that I can reproduce it in just `babel`

**Comment by sebmck** [comment]
I don't think I rebuilt the website for 5.4.5, I think it's on 5.4.4.

**Comment by job13er** [comment]
Here's the whole file:
```js
/**
 * @author: Peter Banka <peter.banka@cyaninc.com>
 * @copyright 2014, Cyan, Inc. All rights reserved.
 */

'use strict';

const $ = require('jquery');
const _ = require('lodash');
const React = require('react');

const menuLinkClasses = require('./menu-link-classes');
const MenuLink = menuLinkClasses.MenuLink;
const panelUtils = require('./panel-utils');

const MOUNT_CHECK_INTERVAL = 50;
const UPDATE_CHECK_INTERVAL = 2;
const NAVBAR_HEIGHT = 34;

/**
 * Check if the given list of menu items has the active one
 * @param {MenuItem[]} menuItems - the items to check
 * @returns {Boolean} true if one of them is active
 */
function hasActiveMenu(menuItems) {
    for (let i = 0, len = menuItems.length; i < len; i++) {
        let menuItem = menuItems[i];
        if (panelUtils.determineActiveMenu(menuItem.idName)) {
            return true;
        }

        if (menuItem.subMenus && hasActiveMenu(menuItem.subMenus)) {
            return true;
        }

    }

    return false;
}

/**
 * Handles navbar items that don't fit into the navbar my squirreling them
 * away into the "menu" dropdown
 */
const NavDropDown = React.createClass({
    displayName: 'NavDropDown',

    propTypes: {
        id: React.PropTypes.string,
        label: React.PropTypes.string,
        menuItems: React.PropTypes.array,
    },

    /**
     * obvious
     * @returns {ReactElement} - content
     */
    render: function () {

        let classNames = 'dropdown';
        if (hasActiveMenu(this.props.menuItems)) {
            classNames += ' active';
        }

        const menuItemComponents = _.map(this.props.menuItems, menuItem => {
            return (
                <li>
                    <MenuLink menuItem={menuItem}/>
                </li>
            );
        });

        return (
            <li className={classNames} id={this.props.id}>
                <a className="dropdown-toggle" data-toggle="dropdown" href="#">
                    {this.props.label}
                    <span className="caret"/>
                </a>
                <ul className="dropdown-menu dropdown-menu-expanded" role="menu">
                    {menuItemComponents}
                </ul>
            </li>
        );
    },
});

/**
 * Handles navbar items that don't fit into the navbar my squirreling them
 * away into the "menu" dropdown
 */
const NavBarOverflow = React.createClass({
    displayName: 'NavBarOverflow',

    propTypes: {
        enabled: React.PropTypes.object, // TODO: make more specific
        hiddenMenuCount: React.PropTypes.number,
        menuLabel: React.PropTypes.string,
        menuList: React.PropTypes.array,
    },

    /**
     * Determine which menu items are hidden (squirreled away)
     * @returns {MenuItem[]} - list of MenuItems
     */
    getHiddenMenuItems: function () {
        const hiddenMenuItems = [];
        const start = this.props.menuList.length - this.props.hiddenMenuCount;
        for (let i = start, len = this.props.menuList.length; i < len; i++) {
            const menuItem = this.props.menuList[i];
            if (!this.props.enabled[menuItem.idName]) {
                continue;
            }

            if (menuItem.subMenus) {
                for (let j = 0, jlen = menuItem.subMenus.length; j < jlen; j++) {
                    let subMenuItem = menuItem.subMenus[j];
                    let newSubMenuItem = _.clone(subMenuItem);
                    newSubMenuItem.name = `${menuItem.name} - ${subMenuItem.name}`;
                    hiddenMenuItems.push(newSubMenuItem);
                }
            } else {
                hiddenMenuItems.push(menuItem);
            }
        }

        return hiddenMenuItems;
    },

    /**
     * obvious
     * @returns {ReactElement} - content
     */

    render: function () {
        if (this.props.hiddenMenuCount === 0) {
            return null;
        }
        const hiddenMenuItems = this.getHiddenMenuItems();
        const hiddenMenuList = this.props.menuList.slice(-1 * this.props.hiddenMenuCount);
        const activeHidden = hasActiveMenu(hiddenMenuList);

        let classNames = 'dropdown';
        if (activeHidden) {
            classNames += ' active';
        }

        return (
            <NavDropDown
                id="cylayout-menu-dropdown"
                label={this.props.menuLabel}
                menuItems={hiddenMenuItems}
            />
        );
    },
});

/**
 * One of the primary menu items on the navbar
 */
const NavItemClass = React.createClass({

    displayName: 'NavItemClass',

    propTypes: {
        defaultActive: React.PropTypes.bool,
        enabled: React.PropTypes.bool,
        hidden: React.PropTypes.bool,
        menuItem: React.PropTypes.object, // TODO: make more specific
    },

    /**
     * Responsible for determining if the menuItem is "active" (i.e. is currently
     * selected)
     * @returns {String} - either 'active' or ''
     */
    getClassName: function () {
        let className;
        const activeLink = panelUtils.determineActiveMenu(this.props.menuItem.idName);
        if (activeLink === undefined) {
            className = this.props.defaultActive ? 'active' : '';
        } else {
            className = activeLink ? 'active' : '';
        }
        return className;
    },

    /**
     * obvious
     * @returns {ReactElement} - content
     */
    render: function () {
        let display = 'block';
        const active = this.getClassName();
        if (!this.props.enabled || this.props.hidden) {
            display = 'none';
        }
        return (
            <li className={active} style={{display: display}}>
                <MenuLink menuItem={this.props.menuItem}/>
            </li>
        );
    },
});

/**
 * this is the little pipe symbol that only shows up if the menu
 * dropdown is being displayed.
 */
const MaybeSeparator = React.createClass({
    displayName: 'MaybeSeparator',

    propTypes: {
        hiddenMenuCount: React.PropTypes.number,
    },

    /**
     * obvious
     * @returns {ReactElement} - content
     */
    render: function () {
        if (this.props.hiddenMenuCount === 0) {
            return null;
        }
        return (
            <li className="pipe-separator" id="cylayout-menu-dropdown-separator" />
        );
    },
});

/**
 * responsible for dyanmically determining how many items are shown and how many
 * items are stuffed into the menu dropdown.
 */
const NavBarLeftMenu = React.createClass({
    displayName: 'NavBarLeftMenu',

    propTypes: {
        enabled: React.PropTypes.object, // TODO: make more specific
        initialActive: React.PropTypes.number,
        menuLabel: React.PropTypes.string,
        menuList: React.PropTypes.array,
        startDeferred: React.PropTypes.object, // TODO: make more specific
        stopDeferred: React.PropTypes.object, // TODO: make more specific
    },

    /**
     * Keeps track of internal details about what menu items are being
     * displayed at the current time due to viewport-width pressure
     * @returns {StateObject} - the initial state
     */
    getInitialState: function () {
        return {
            hiddenMenuCount: 0,
            // keeps track of whether we have reached the maximum number
            // of menu items and we're trying to reduce-to-fit. Prevents
            // layout widget from speculatively adding more nav items in the
            // hopes of showing the user more.
            shrinking: false,
            // Only want to create the resize hook once
            createdResizeHook: false,
            // We only want one setTimeout action going at once
            timeoutId: null,
        };
    },

    /**
     * Determine if there is space left in the header
     * @returns {Boolean} true if there is some space left
     */
    _spaceLeft: function () {
        const $container = $('#cylayout-nav-container');
        // NOTE: if there are no items in the navbar, it can
        //       be *smaller* than its parent
        return $container.parent().height() <= NAVBAR_HEIGHT;
    },

    /**
     * If there is not enough room, disable a menu item
     */
    _disableOneMenuItem: function () {
        if (this.state.hiddenMenuCount < this.props.menuList.length) {
            const newHiddenMenuCount = this.state.hiddenMenuCount + 1;
            this.setState({hiddenMenuCount: newHiddenMenuCount});
        }
    },

    /**
     * If there might be more room, enable a menu item
     */
    _enableOneMenuItem: function () {
        if (this.state.hiddenMenuCount > 0) {
            const newHiddenMenuCount = this.state.hiddenMenuCount - 1;
            this.setState({hiddenMenuCount: newHiddenMenuCount});
        } else {
            this._steadyState();
        }
    },

    _steadyState: function () {
        this.shrinking = false; // stable-state
        // console.log('resolving start deferred!');
        this.props.startDeferred.resolve();
    },

    /**
     * Determine if navitems need to be added or removed
     */
    _checkYourself: function () {
        this.timeoutId = null;
        const spaceLeft = this._spaceLeft();
        const growing = spaceLeft && !this.shrinking;
        if (!spaceLeft) {
            this.shrinking = true;
            this._disableOneMenuItem();
        } else {
            if (growing) {
                this._enableOneMenuItem();
            } else {
                this._steadyState();
            }
        }
    },

    /**
     * queue up a call _checkYourself() and create resize hook
     */
    componentDidMount: function () {
        if (!this.timeoutId) {
            this.timeoutId = window.setTimeout(() => this._checkYourself(), MOUNT_CHECK_INTERVAL);
        }
        if (!this.state.createdResizeHook) {
            $(window).on('resize.layout-widget', () => {
                try {
                    this.forceUpdate(); // May want to do this less often
                } catch (e) {
                    // Can raise if the object is being torn down
                    //console.log('NavBarLeftMenu: ', e);
                }
            });
            this.setState({'createdResizeHook': true});
        }
    },

    /**
     * Component just updated (likely because a nav-item was just added or removed.
     * queue up a call to _checkYourself()
     */
    componentDidUpdate: function () {
        if (!this.timeoutId) {
            this.timeoutId = window.setTimeout(() => this._checkYourself(), UPDATE_CHECK_INTERVAL);
        }
    },

    /**
     * obvious
     * @returns {ReactElement} - content
     */
    render: function () {
        const navItems = [];
        const hiddenIndex = this.props.menuList.length - this.state.hiddenMenuCount;
        // If only one item, it's enabled and hidden
        if (this.props.menuList.length === 1) {
            return (
                <ul className="nav nav-pills" id="cylayout-nav-container">
                    <NavItemClass
                        defaultActive={true}
                        enabled={true}
                        hidden={true}
                        key={this.props.menuList[0].idName}
                        menuItem={this.props.menuList[0]}
                    />;
                </ul>
            );
        } else {
            for (let i = 0, len = this.props.menuList.length; i < len; i++) {
                const menuItem = this.props.menuList[i];
                const enabled = this.props.enabled[menuItem.idName];
                const hidden = i >= hiddenIndex;
                const defaultActive = (i === this.props.initialActive) ? true : false;

                if (menuItem.subMenus && !menuItem && menuItem) {
                    navItems.push(
                        <NavDropDown
                            id={menuItem.linkName}
                            label={menuItem.name}
                            menuItems={menuItem.subMenus}
                        />
                    );
                } else {
                    navItems.push(
                        <NavItemClass
                            defaultActive={defaultActive}
                            enabled={enabled}
                            hidden={hidden}
                            key={menuItem.idName}
                            menuItem={menuItem}
                        />
                    );
                }

            }
        }
        return (
            <ul className="nav nav-pills" id="cylayout-nav-container">
                {navItems}
                <MaybeSeparator hiddenMenuCount={this.state.hiddenMenuCount}/>
                <NavBarOverflow
                    enabled={this.props.enabled}
                    hiddenMenuCount={this.state.hiddenMenuCount}
                    menuLabel={this.props.menuLabel}
                    menuList={this.props.menuList}
                />
            </ul>
        );
    },
});

module.exports = NavBarLeftMenu;
```

**Comment by job13er** [comment]
you can see when running it through `babel` that the first `menuItem` becomes `_menuItem`

The problem could be related to us having multiple react components in one file. That particular file is in need of some refactoring, but still seems like a bug to replace the `menuItem={menuItem}` with `_menuItem3: _menuItem3` vs. `menuItem3: _menuItem3` or `menuItem: menuItem` (which is what you get with that code snippet in isolation).

**Comment by job13er** [comment]
Wow, super service! Thanks @sebmck :+1:

**Comment by sebmck** [comment]
Fixed as of 5.4.7, thanks!

**Comment by psbanka** [comment]
You rule :metal: ; thanks for the quick turn-around
